include/h2o/pipe_utils.h

```diff
@@ -0,0 +1,39 @@
+#ifndef __H2O__PIPE_UTILS__H__
+#define __H2O__PIPE_UTILS__H__
+
+#include <stddef.h>
+#include <stdbool.h>
+
+enum h2o_send_state;
+struct st_h2o_req_t;
+struct st_h2o_context_t;
+
+typedef struct st_h2o_pipe_reader_t {
+    int fds[2]; /* fd[0] set to -1 unless used */
+    int inflight;
+
+    size_t body_bytes_read, body_bytes_sent;
+} h2o_pipe_reader_t;
+
+int h2o_empty_pipe(int fd);
+
+void h2o_pipe_reader_send(struct st_h2o_req_t *req, h2o_pipe_reader_t *pipe_reader, enum h2o_send_state send_state);
```

Comment:
I am using this declaration as an example, but in general, I would not mind calling the 2nd argument "reader", because maybe that is obvious. Alongside the inclusion of h2o.h (see my comment above), the declaration would then look like:
```suggestion
void h2o_pipe_reader_send(h2o_req_t *req, h2o_pipe_reader_t *reader, h2o_send_state_t send_state);
```